MoltenVK/MoltenVK/Commands/MVKMTLBufferAllocation.mm

```diff
@@ -17,14 +17,20 @@
  */
 
 #include "MVKMTLBufferAllocation.h"
+#include <sys/mman.h>
 
 
 #pragma mark -
 #pragma mark MVKMTLBufferAllocation
 
 MVKVulkanAPIObject* MVKMTLBufferAllocation::getVulkanAPIObject() { return _pool->getVulkanAPIObject(); };
 
-void MVKMTLBufferAllocation::returnToPool() const { _pool->returnAllocation((MVKMTLBufferAllocation*)this); }
+void MVKMTLBufferAllocation::returnToPool() {
+    std::lock_guard<std::mutex> lock(_pool->_lock);
```

Comment:
I'd prefer to leave `MVKObjectPool::_lock` protected, and move this operation to `MVKMTLBufferAllocationPool`, similar to the way `MVKMTLBufferAllocationPool:: returnAllocation()` was designed. That might be easier if we are able to revert to using `mlock()/munlock()` on each allocation range during acquisition and return to pool, as I've suggested.

Also, you seem to be assuming that `_isThreadSafe` is always enabled, and the mutex is always needed. It would be better to honor  the state of `_isThreadSafe`.

From what I can see, it looks like all our internal use cases actually do always enable `_isThreadSafe`, so if it is too challenging to design this class to allow `_isThreadSafe` to be honored, perhaps a Plan B would be to remove `_isThreadSafe` as an option, and have this class always behave in a threadsafe manner.

Reply:
> I'd prefer to leave `MVKObjectPool::_lock` protected, and move this operation to `MVKMTLBufferAllocationPool`, similar to the way `MVKMTLBufferAllocationPool:: returnAllocation()` was designed. That might be easier if we are able to revert to using `mlock()/munlock()` on each allocation range during acquisition and return to pool, as I've suggested.

OK, I'll look into this more when we agree on the general scheme.

> Also, you seem to be assuming that `_isThreadSafe` is always enabled, and the mutex is always needed. It would be better to honor the state of `_isThreadSafe`.

My bad, I was basing on an older revision where `returnToPool` always called `returnObjectSafely`, and didn't notice that it was changed when rebasing.